src/ngraph_encapsulate_clusters.cc

```diff
@@ -506,7 +507,105 @@ Status EncapsulateClusters(Graph* graph, int graph_id,
+                         [&key](std::set<UniqueTensorId> group) {
+                           return group.find(key) != group.end();
+                         });
+    };
+
+    auto is_encapsulate = [](Node* n) {
+      return n->type_string() == "NGraphEncapsulate";
+    };
+
+    auto get_string_info = [](int enc_id, bool is_output, int slot_id) {
+      std::vector<int> info = {enc_id, is_output ? 1 : 0, slot_id};
+      return ng::join(info, "_");
+    };
+
+    auto add_to_group = [&get_string_info, &is_encapsulate](
+        Node* n, bool is_out, int slot, std::set<UniqueTensorId>& new_group) {
+      if (is_encapsulate(n)) {
+        int enc_idx;
+        TF_RETURN_IF_ERROR(GetNodeAttr(n->attrs(), "ngraph_cluster", &enc_idx));
+        UniqueTensorId tid = get_string_info(enc_idx, is_out, slot);
+        new_group.insert(tid);
+      }
+      return Status::OK();
+    };
+
+    // Main loop for this pass starts here
+    for (auto node : graph->nodes()) {
+      if (is_encapsulate(node)) {
+        int curr_encapsulate_idx;
+        TF_RETURN_IF_ERROR(GetNodeAttr(node->attrs(), "ngraph_cluster",
+                                       &curr_encapsulate_idx));
+        for (auto in_neighbour_edge : node->in_edges()) {
+          Node* in_neighbour = in_neighbour_edge->src();
+          UniqueTensorId curr_tid = get_string_info(
+              curr_encapsulate_idx, false, in_neighbour_edge->dst_input());
+          std::set<UniqueTensorId> new_group{};
+          // TODO: instead of searching in the vector, maybe mark graph nodes
+          // with a "visited" tag
+          if (!is_tracked(curr_tid)) {
+            // Note we have to only run "is_tracked" once.
+            // In this design, if a tensor of a group is inserted in
+            // shared_tensors, then all members of the group are already in
+            // On the flip side if one tensor is not in, no other of the group
+            // is in yet
+            Node* in_neighbour_out = nullptr;
+            int in_neighbour_out_dst_slot, in_neighbour_out_src_slot;
+            for (auto in_neighbour_out_edge : in_neighbour->out_edges()) {
+              in_neighbour_out_dst_slot = in_neighbour_out_edge->dst_input();
+              in_neighbour_out = in_neighbour_out_edge->dst();
+              if (in_neighbour_out_dst_slot >= 0) {  // Skip control edge (-1)
+                if (in_neighbour_out_edge->src_output() ==
+                    in_neighbour_edge->src_output()) {
+                  in_neighbour_out_src_slot =
+                      in_neighbour_out_edge->src_output();
+                  // Add the source of the shareable group
+                  add_to_group(in_neighbour, true, in_neighbour_out_src_slot,
+                               new_group);
+                }
+                // Add the sinks of the shareable group
+                add_to_group(in_neighbour_out, false, in_neighbour_out_dst_slot,
```

Comment:
The sink should only be added to the group if the 
```
in_neighbour_out_edge->src_output() ==
                    in_neighbour_edge->src_output()) 
```
ie the sink should be also part of the if? no?

Reply:
For a graph like below, it seems E2 and E3 are put in the same shared group, when they should not be
```
E1 (output index 1) -----> E2
(output index 0)
|
|
E3
```

Reply:
Will check this, and add a UT. Thanks

Reply:
Would this pseudo-algo be simpler?
```
 For each node:
          create new_grp
          Look at all output edges with src_output_index, i
                     if dst node is enc
                              Add the dstNode_input_id to the new_grp
         if node is enc
                    Add the node_output_id to the new_grp
         if new_grp is trivial 
                   Discard it
```